packages/@lwc/ssr-runtime/src/wire.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { LightningElement } from './lightning-element';
+import type {
+    WireAdapterConstructor,
+    WireContextConsumer,
+    WireContextProviderOptions,
+} from '@lwc/engine-core';
+
+type SsrContextProvider = (le: LightningElement, options?: WireContextProviderOptions) => void;
+
+const contextfulRelationships = new WeakMap<LightningElement, LightningElement>();
+export function establishContextfulRelationship(
+    parentLe: LightningElement,
+    childLe: LightningElement
+): void {
+    contextfulRelationships.set(childLe, parentLe);
+}
```

Comment:
This might be an optimization opportunity. Creating & preserving an actual stack of components is probably a worse solution, as it would 1) be slow, and 2) require passing that stack object through many intermediate functions.

However, I could see doing something like `childLe[SECRET_PARENT_SYMBOL] = parentLe` or something like that rather than using the `WeakMap`. I'm not sure which is faster; I just picked one and went with it.